File: android/src/main/java/com/margelo/nitro/splash/HybridSplashScreen.kt
package com.margelo.nitro.splash

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise

/** Nitro adapter over [SplashScreenManager]. Methods run on the JS thread. */
@DoNotStrip
@Keep
class HybridSplashScreen : HybridSplashScreenSpec() {
  override val manifest: SplashManifestSpec
    get() = SplashScreenManager.manifest

  override val launchInfo: SplashLaunchInfoSpec
    get() = SplashScreenManager.launchInfo()

  override fun isVisible(): Boolean = SplashScreenManager.isVisible

  override fun preventAutoHide() = SplashScreenManager.preventAutoHide()

  override fun hide(options: SplashHideOptionsSpec): Promise<Unit> {
    val promise = Promise<Unit>()
    SplashScreenManager.hide(options.fade, options.durationMs.toLong()) { promise.resolve(Unit) }
    return promise
  }

  override fun show(): Promise<Unit> {
    val promise = Promise<Unit>()
    SplashScreenManager.show { promise.resolve(Unit) }
    return promise
  }

  override fun setEventListener(listener: (event: SplashNativeEvent) -> Unit) {
    SplashScreenManager.eventListener = listener
  }

  override fun clearEventListener() {
    SplashScreenManager.eventListener = null
  }
}

File: ios/HybridSplashScreen.swift
import Foundation
import NitroModules

/// Nitro adapter over `SplashScreenManager`. Methods run on the JS thread;
/// the manager hops to the main thread where UIKit needs it.
final class HybridSplashScreen: HybridSplashScreenSpec {
  private let manager = SplashScreenManager.shared

  var manifest: SplashManifestSpec {
    manager.manifest
  }

  var launchInfo: SplashLaunchInfoSpec {
    manager.launchInfo()
  }

  func isVisible() throws -> Bool {
    manager.isVisible
  }

  func preventAutoHide() throws {
    manager.preventAutoHide()
  }

  func hide(options: SplashHideOptionsSpec) throws -> Promise<Void> {
    let promise = Promise<Void>()
    manager.hide(fade: options.fade, durationMs: options.durationMs) {
      promise.resolve()
    }
    return promise
  }

  func show() throws -> Promise<Void> {
    let promise = Promise<Void>()
    manager.show {
      promise.resolve()
    }
    return promise
  }

  func setEventListener(listener: @escaping (_ event: SplashNativeEvent) -> Void) throws {
    manager.eventListener = listener
  }

  func clearEventListener() throws {
    manager.eventListener = nil
  }
}
